streaming/src/test/scala/org/apache/spark/streaming/LocalStreamingContext.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming
+
+import org.scalatest.{BeforeAndAfterAll, BeforeAndAfterEach, Suite}
+
+import org.apache.spark.SparkContext
+import org.apache.spark.internal.Logging
+
+/**
+ * Manages a local `ssc` `StreamingContext` variable, correctly stopping it after each test.
+ * Note that it also stops active SparkContext if `stopSparkContext` is set to true (default).
+ * In most cases you may want to leave it, to isolate environment for SparkContext in each test.
+ */
+trait LocalStreamingContext extends BeforeAndAfterEach with BeforeAndAfterAll { self: Suite =>
+
+  @transient var ssc: StreamingContext = _
+  @transient var stopSparkContext: Boolean = true
+
+  override def beforeAll() {
+    super.beforeAll()
+  }
+
+  override def afterEach() {
+    try {
+      resetStreamingContext()
+    } finally {
+      super.afterEach()
+    }
+  }
+
+  def resetStreamingContext(): Unit = {
+    LocalStreamingContext.stop(ssc, stopSparkContext)
+    ssc = null
+  }
+}
+
+object LocalStreamingContext extends Logging {
+  def stop(ssc: StreamingContext, stopSparkContext: Boolean): Unit = {
+    if (stopSparkContext) {
+      ensureNoActiveSparkContext(ssc)
+    } else if (ssc != null) {
+      ssc.stop(stopSparkContext = false)
+    }
+  }
+
+  /**
+   * Clean up active SparkContext: try to stop first if there's an active SparkContext.
+   * If it fails to stop, log warning message and clear active SparkContext to avoid
+   * interfere between tests.
+   */
+  def ensureNoActiveSparkContext(): Unit = {
+    // if SparkContext is still active, try to clean up
+    SparkContext.getActive match {
+      case Some(sc) =>
+        try {
+          sc.stop()
+        } catch {
+          case e: Throwable =>
+            logWarning("Exception trying to stop SparkContext, clear active SparkContext...", e)
+            SparkContext.clearActiveContext()
+        }
+      case _ =>
+    }
+  }
+
+  def ensureNoActiveSparkContext(ssc: StreamingContext): Unit = {
```

Comment:
This function does a little bit more + I have the feeling this can be merged with `stop`.
```
  def stop(ssc: StreamingContext, stopSparkContext: Boolean): Unit = {
    if (ssc != null) {
      ssc.stop(stopSparkContext)
    }
    ensureNoActiveSparkContext()
  }
```
BTW, why are these functions are belonging to the object? There are transient vars for `ssc` and `stopSparkContext`.

Reply:
I'll inline this to `stop()` as there's no usage other than `stop()`.

Regarding placing them to the object instead of trait, it can be also used from outside of LocalStreamingContext trait. `LocalSparkContext` also exposes similar thing. I'll modify the code shortly - it will help deduplicate the code outside of `LocalSparkContext`.